test/provisioner/provision_ibmcloud.go

```diff
@@ -5,33 +5,455 @@ package provisioner
+		return err
+	}
+
+	// TODO, wait cluter ready, it takes long time
 	return nil
 }
 
-func (l *IBMCloudProvisioner) CreateVPC(ctx context.Context, cfg *envconf.Config) error {
-	return nil
+func (p *IBMCloudProvisioner) CreateVPC(ctx context.Context, cfg *envconf.Config) error {
+	ibmcloudTrace("CreateVPC()")
+	return createVpcImpl()
 }
 
-func (l *IBMCloudProvisioner) DeleteCluster(ctx context.Context, cfg *envconf.Config) error {
+func (p *IBMCloudProvisioner) DeleteCluster(ctx context.Context, cfg *envconf.Config) error {
+	ibmcloudTrace("DeleteCluster()")
+
+	target := containerv2.ClusterTargetHeader{}
+	ibmcloudTrace(fmt.Sprintf("Deleting Cluster %s.\n", IBMCloudProps.ClusterName))
+	err := IBMCloudProps.ClusterAPI.Delete(IBMCloudProps.ClusterName, target)
+	if err != nil {
+		return err
+	}
+
+	ibmcloudTrace(fmt.Sprintf("Deleted Cluster %s.\n", IBMCloudProps.ClusterName))
 	return nil
 }
 
-func (l *IBMCloudProvisioner) DeleteVPC(ctx context.Context, cfg *envconf.Config) error {
-	return nil
+func (p *IBMCloudProvisioner) DeleteVPC(ctx context.Context, cfg *envconf.Config) error {
+	ibmcloudTrace("DeleteVPC()")
+	return deleteVpcImpl()
 }
 
-func (l *IBMCloudProvisioner) UploadPodvm(imagePath string, ctx context.Context, cfg *envconf.Config) error {
+func (p *IBMCloudProvisioner) UploadPodvm(imagePath string, ctx context.Context, cfg *envconf.Config) error {
+	ibmcloudTrace("UploadPodvm()")
+
+	filePath, err := filepath.Abs(imagePath)
+	if err != nil {
+		return err
+	}
+
+	conf := aws.NewConfig().
+		WithEndpoint(IBMCloudProps.CosServiceURL).
+		WithCredentials(ibmiam.NewStaticCredentials(aws.NewConfig(),
+			IBMCloudProps.IamServiceURL, IBMCloudProps.CosApiKey, IBMCloudProps.CosInstanceID)).
+		WithS3ForcePathStyle(true)
+
+	sess := cosession.Must(cosession.NewSession(conf))
+	ibmcloudTrace("session initialized.")
+
+	file, err := os.Open(imagePath)
+	if err != nil {
+		return err
+	}
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+	ibmcloudTrace(fmt.Sprintf("qcow2 image file %s validated.\n", imagePath))
+
+	reader := &utils.CustomReader{
+		Fp:      file,
+		Size:    fileInfo.Size(),
+		SignMap: map[int64]struct{}{},
+	}
+
+	uploader := s3manager.NewUploader(sess, func(u *s3manager.Uploader) {
+		u.PartSize = 5 * 1024 * 1024
+		u.LeavePartsOnError = true
+	})
+
+	key := filepath.Base(filePath)
+	_, err = uploader.Upload(&s3manager.UploadInput{
+		Bucket: aws.String(IBMCloudProps.Bucket),
+		Key:    aws.String(key),
+		Body:   reader,
+	})
+	if err != nil {
+		return err
+	}
+	ibmcloudTrace(fmt.Sprintf("File %s uploaded to bucket.\n", key))
+
+	var osNames []string
+	if strings.EqualFold("s390x", IBMCloudProps.PodvmImageArch) {
+		osNames = []string{"ubuntu-20-04-s390x"}
+	} else {
+		osNames = []string{"ubuntu-20-04-amd64"}
+
+	}
+	operatingSystemIdentityModel := &vpcv1.OperatingSystemIdentityByName{
+		Name: &osNames[0],
+	}
+
+	cosID := "cos://" + IBMCloudProps.Region + "/" + IBMCloudProps.Bucket + "/" + key
```

Comment:
This might not be supported yet, but I tried to upload to a cos bucket in ca-tor (as I already had one created there), but for a cluster to be created in jp-tok (to keep my demo environment in ca-tor clean), and this failed as `IBMCloudProps.Region` is pointing to `jp-tok`, so this produced an id of `cos://jp-tok/sh-pp-z-cos-bucket-ca-tor-2-2-23/podvm.qcow2`, but my cos id is: `cos://ca-tor/sh-pp-z-cos-bucket-ca-tor-2-2-23/podvm.qcow2`.

Reply:
It addition to this point there is a bit of a disconnect with `VPC_SERVICE_URL` and the region that we might need to reconcile. After the above issue I changed `REGION="ca-tor"` and `ZONE="ca-tor-2"`, but forgot to update `VPC_SERVICE_URL` which resulted in the upload working:
```
total read:702414848    progress:100%File podvm.qcow2 uploaded to bucket.

cosID cos://ca-tor/sh-pp-z-cos-bucket-ca-tor-2-2-23/podvm.qcow2, imageName podvm.

Image podvm.qcow2 with PodvmImageID r022-7c0f2181-a8b5-45e2-b02f-daef85c28534 created from the bucket.
```
but it was actually created in jp-tok rather confusingly!
```
$ ibmcloud is images --visibility=private
Listing images in all resource groups and region jp-tok under account Steven Horsman's Account as user steven@uk.ibm.com...
ID                                          Name    Status      Arch    OS name              OS version                              File size(GB)   Visibility   Owner type   Encryption   Resource group   Catalog Offering
r022-7c0f2181-a8b5-45e2-b02f-daef85c28534   podvm   available   s390x   ubuntu-20-04-s390x   20.04 LTS Focal Fossa Minimal Install   1               private      user         none         default
```

Reply:
I'm not sure whether it's support that the cos and cluster are in different region, assumption is not. Will do some investigating later.